cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1024 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/columnar_format.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/array/builder_nested.h"
+#include "arrow/array/builder_binary.h"
+#include <iostream>
+
+
+namespace arrow {
+
+// ColumnMap implementation
+
+class ColumnMap::Impl {
+ public:
+  class ColumnData {
+   public:
+    std::shared_ptr<Field> field;
+    std::shared_ptr<Int16Array> rep_levels;
+    std::shared_ptr<Int16Array> def_levels;
+    std::shared_ptr<Array> values;
+  };
+
+  // Return position if found or -(ins_pos-1) if not found
+  int find(const std::shared_ptr<Field>& field) const;
+
+  // Elements are ordered by pointer value ColumnData::field::get()
+  std::vector<ColumnData> columns_;
+};
+
+int ColumnMap::Impl::find(const std::shared_ptr<Field>& field) const {
+  // First, binary search is used (equality is by pointer value).
+  // This helps to locate elements fast when the objects from the schema are queried for.
+  int low = 0;
```

Comment:
Could you just use `std::lower_bound`?

Reply:
I tried. It makes code harder to read (due to iterator differences and lambdas).